src/services/organizationService.ts

```diff
@@ -490,6 +495,104 @@ export async function registerOrganization(
   }
 }
 
+export function addDelegate(
+  ipaCode: string,
+  userEmail: string
+): TaskEither<
+  // tslint:disable-next-line:max-union-size
+  IResponseErrorInternal | IResponseErrorNotFound | IResponseErrorConflict,
+  IResponseSuccessRedirectToResource<Organization, Organization>
+> {
+  const genericErrorHandler = (error: unknown) => {
+    log.error("An error occurred adding the delegate. %s", error);
+    return ResponseErrorInternal("An error occurred adding the delegate.");
+  };
+  return tryCatch<
+    IResponseErrorInternal | IResponseErrorNotFound | IResponseErrorConflict,
+    OrganizationModel
+  >(
+    () =>
+      OrganizationModel.findOne({
+        where: { ipaCode }
+      }),
+    genericErrorHandler
+  )
+    .chain(
+      fromPredicate<IResponseErrorNotFound, OrganizationModel>(
+        _ => _ !== null,
+        () =>
+          ResponseErrorNotFound(
+            "Not found",
+            "Registered organization does not exists"
+          )
+      )
+    )
+    .chain(
+      fromPredicate<
+        IResponseErrorConflict | IResponseErrorInternal,
+        OrganizationModel
+      >(
+        _ =>
+          _.registrationStatus ===
+          OrganizationRegistrationStatusEnum.REGISTERED,
+        () =>
+          ResponseErrorConflict(
+            "The organization has not completed the registration process yet."
+          )
+      )
+    )
+    .chain(organizationModel =>
+      tryCatch<IResponseErrorInternal, OrganizationModel>(
+        () =>
+          OrganizationUserModel.create({
+            createdAt: Date.now(),
+            organizationIpaCode: organizationModel.ipaCode,
+            updatedAt: Date.now(),
+            userEmail,
+            userRole: UserRoleEnum.ORG_DELEGATE
+          }).then(_ => organizationModel),
```

Comment:
I haven't checked but maybe you can omit this `.then(...)` if you use `chainFirst()`

Reply:
I don't understand how I should refactor the code in order to use the `chainFirst()` method...

Reply:
nevermind